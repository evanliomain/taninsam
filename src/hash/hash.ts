/**
 * @module any=>string
 */
import { isBoolean } from '../isBoolean';
import { isString } from '../isString';
import { isNumber } from '../isNumber';
import { isArray } from '../isArray';
import { isObject } from '../isObject';
import { entries } from '../entries';
import { chain } from '../chain';
import { map } from '../map';
import { sortBy } from '../sortBy';
import { join } from '../join';
import { isNull } from '../isNull';
import { isUndefined } from '../isUndefined';
/**
 * Convert the passed value into a string.
 * 2 values that have the same "value" but are referenced
 * with 2 variables will have the same hash.
 * @return the function to apply on the TBD to do something
 * @example
 * ```
 * hash()(1) === hash()(1)
 * hash()(1) !== hash()("1")
 * hash()(true) !== hash()("true")
 * ```
 */
export function hash<T>(): (value: T) => string {
  return (value: T) => {
    if (isNull(value)) {
      return 'null';
    }
    if (isUndefined(value)) {
      return 'undefined';
    }
    if (isBoolean(value)) {
      return `b(${value.toString()})`;
    }
    if (isString(value)) {
      return `s(${value})`;
    }
    if (isNumber(value)) {
      return `n(${value.toString()})`;
    }
    if (isArray(value)) {
      return `a([${value.map(element => hash()(element)).join(',')}])`;
    }
    if (isObject(value)) {
      return chain(value)
        .chain(entries())
        .chain(sortBy(([key, _]) => key))
        .chain(map(([key, v]) => `${key}:${hash()(v)}`))
        .chain(join(','))
        .chain(s => `o({${s}})`)
        .value();
    }
    throw new Error(`Try to hash an unsupported value type: ${typeof value}`);
  };
}
