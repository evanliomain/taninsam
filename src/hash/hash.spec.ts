import { describe, expect, test } from 'vitest';
import { hash } from './hash';

describe('hash function', () => {
  const input = 'something';

  test('is a pure function', () => {
    hash()(input);
    expect(input).toMatchSnapshot();
  });

  test('hash 1 === hash 1', () => {
    expect(compare(1, 1)).toBeTruthy();
  });

  test('hash 1 !== hash 2', () => {
    expect(compare(1, 2)).toBeFalsy();
  });

  test('hash 1 !== hash 2', () => {
    expect(compare(1, 2)).toBeFalsy();
  });

  test('hash true !== hash true', () => {
    expect(compare(true, true)).toBeTruthy();
  });

  test('hash true !== hash false', () => {
    expect(compare(true, false)).toBeFalsy();
  });

  test('hash true !== hash "true"', () => {
    expect(compare(true, 'true')).toBeFalsy();
  });

  test('hash false !== hash "false"', () => {
    expect(compare(false, 'false')).toBeFalsy();
  });

  test('hash 1 !== hash "1"', () => {
    expect(compare(1, '1')).toBeFalsy();
  });

  test('hash { a: 1, b: 2 } === hash { b: 2, a: 1 }', () => {
    expect(compare({ a: 1, b: 2 }, { b: 2, a: 1 })).toBeTruthy();
  });

  test('hash { a: 1, b: 2 } !== hash { a: 1, b: 2, c: 3 }', () => {
    expect(compare({ a: 1, b: 2 }, { a: 1, b: 2, c: 3 })).toBeFalsy();
  });

  test('hash [1, 2] === hash [1, 2]', () => {
    expect(compare([1, 2], [1, 2])).toBeTruthy();
  });

  test('hash [1, 2] !== hash [2, 1]', () => {
    expect(compare([1, 2], [2, 1])).toBeFalsy();
  });

  test('hash [{ a: 1, b: 2 }, { a: 3, b: 4 }] === hash [{ b: 2, a: 1 }, { a: 3, b: 4 }]', () => {
    expect(
      compare(
        [
          { a: 1, b: 2 },
          { a: 3, b: 4 },
        ],
        [
          { b: 2, a: 1 },
          { a: 3, b: 4 },
        ],
      ),
    ).toBeTruthy();
  });

  test('hash [{ a: 1, b: 2 }, { a: 3, b: 4 }] !== hash [{ a: 3, b: 4 }, { a: 1, b: 2 }]', () => {
    expect(
      compare(
        [
          { a: 1, b: 2 },
          { a: 3, b: 4 },
        ],
        [
          { a: 3, b: 4 },
          { a: 1, b: 2 },
        ],
      ),
    ).toBeFalsy();
  });

  test('hash undefined !== hash null', () => {
    expect(compare(undefined, null)).toBeFalsy();
  });

  test('hash undefined !== hash "undefined"', () => {
    expect(compare(undefined, 'undefined')).toBeFalsy();
  });

  test('hash null !== hash "null"', () => {
    expect(compare(null, 'null')).toBeFalsy();
  });

  test('hash undefined !== hash 0', () => {
    expect(compare(undefined, 0)).toBeFalsy();
  });

  test('hash null !== hash 0', () => {
    expect(compare(null, 0)).toBeFalsy();
  });

  test('hash({ f: () => 1 }) throw exception', () => {
    expect(() => hash()({ f: () => 1 })).toThrowErrorMatchingSnapshot();
  });

  function compare(a, b): boolean {
    return hash()(a) === hash()(b);
  }
});
