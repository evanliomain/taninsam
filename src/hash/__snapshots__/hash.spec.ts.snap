// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash function hash({ f: () => 1 }) throw exception 1`] = `"Try to hash an unsupported value type: function"`;

exports[`hash function is a pure function 1`] = `"something"`;
